CMakeLists.txt

```diff
@@ -0,0 +1,164 @@
+
+target_sources(deflate PRIVATE ${LIBDEFLATE_SOURCES}
+                               ${LIBDEFLATE_HEADERS})
+
+# Compile options
+check_c_compiler_flag(-Wall DEFLATE_LINT_WALL)
+check_c_compiler_flag(-Wundef DEFLATE_LINT_WUNDEF)
+check_c_compiler_flag(-Wpedantic DEFLATE_LINT_PEDANTIC)
+check_c_compiler_flag(-Wdeclaration-after-statement DEFLATE_LINT_DECLARATION_AFTER_STATEMENT)
+check_c_compiler_flag(-Wmissing-prototypes DEFLATE_LINT_MISSING_PROTOTYPES)
+check_c_compiler_flag(-Wstrict-prototypes DEFLATE_LINT_STRICT_PROTOTYPES)
+check_c_compiler_flag(-Wvla DEFLATE_LINT_VLA)
+check_c_compiler_flag(-Wimplicit-fallthrough DEFLATE_LINT_IMPLICIT_FALLTHROUGH)
+
+set(lint-all $<$<BOOL:${DEFLATE_LINT_WALL}>:-Wall>)
+set(lint-undef $<$<BOOL:${DEFLATE_LINT_UNDEF}>:-Wundef>)
+set(lint-pedantic $<$<BOOL:${DEFLATE_LINT_PEDANTIC}>:-Wpedantic>)
+set(lint-declaration-after-statement $<$<BOOL:${DEFLATE_LINT_DECLARATION_AFTER_STATEMENT}>:-Wdeclaration-after-statement>)
+set(lint-missing-prototypes $<$<BOOL:${DEFLATE_LINT_MISSING_PROTOTYPES}>:-Wmissing-prototypes>)
+set(lint-strict-prototypes $<$<BOOL:${DEFLATE_LINT_STRICT_PROTOTYPES}>:-Wstrict-prototypes>)
+set(lint-vla $<$<BOOL:${DEFLATE_LINT_VLA}>:-Wvla>)
+set(lint-implicit-fallthrough $<$<BOOL:${DEFLATE_LINT_IMPLICIT_FALLTHROUGH}>:-Wimplicit-fallthrough>)
+
+target_compile_options(deflate PRIVATE ${lint-all})
+target_compile_options(deflate PRIVATE ${lint-undef})
+target_compile_options(deflate PRIVATE ${lint-pedantic})
+target_compile_options(deflate PRIVATE ${lint-declaration-after-statement})
+target_compile_options(deflate PRIVATE ${lint-missing-prototypes})
+target_compile_options(deflate PRIVATE ${lint-strict-prototypes})
+target_compile_options(deflate PRIVATE ${lint-vla})
+target_compile_options(deflate PRIVATE ${lint-implicit-fallthrough})
+
+# The CFLAGS environment variable is read and handled automatically by cmake
+
+if (BUILD_SHARED_LIBS)
+  target_compile_definitions(deflate PRIVATE _ANSI_SOURCE)
+  set_target_properties(deflate PROPERTIES C_VISIBILITY_PRESET hidden)
+
+  if (WIN32)
+    target_compile_definitions(deflate PUBLIC LIBDEFLATE_DLL)
+  endif()
+endif()
+
+if (FREESTANDING)
+  target_compile_definitions(deflate PRIVATE FREESTANDING)
+  target_compile_options(deflate PRIVATE -ffreestanding -nostdlib)
+endif()
+
+if (DEFLATE_TEST_SUPPORT__DO_NOT_USE)
+  target_compile_definitions(deflate PRIVATE TEST_SUPPORT__DO_NOT_USE)
+endif()
+
+target_include_directories(deflate PUBLIC
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+  $<INSTALL_INTERFACE:include>
+)
+
+set_target_properties(deflate PROPERTIES PUBLIC_HEADER "libdeflate.h")
+
+install(TARGETS deflate
+        EXPORT deflate
+        ARCHIVE DESTINATION lib
+        PUBLIC_HEADER DESTINATION include/
+)
+
+# This is required to work with FetchContent
+install(EXPORT deflate
+        FILE deflate-config.cmake
+        NAMESPACE deflate::
+        DESTINATION lib/cmake/deflate)
```

Comment:
- Misleading comment, CMake config files are mainly for `find_package()`.
- Instead I would write a basic `deflate-config.cmake.in` file and rely on `CMakePackageConfigHelpers` module and its `configure_package_config_file`. I would export targets in `deflate-targets.cmake`. Do not forget to also export a target file in the build tree. A version file would be nice also (it requires to add VERSION to the project.
  Basically something like this:

  ```cmake
  configure_package_config_file(
      ${PROJECT_SOURCE_DIR}/cmake/deflate-config.cmake.in
      deflate-config.cmake
      INSTALL_DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/deflate
  )
  
  write_basic_package_version_file(
      deflate-config-version.cmake
      VERSION ${PROJECT_VERSION}
      COMPATIBILITY AnyNewerVersion
  )
  
  export(
      EXPORT deflate
      NAMESPACE deflate::
      FILE deflate-targets.cmake
  )
  
  install(
      EXPORT deflate
      NAMESPACE deflate::
      FILE deflate-targets.cmake
      DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/deflate
  )
  ```

- The makefile installs a pkgconfig file, it should be ported to the CMakeLists. The previous one can be reused.